python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -132,7 +156,61 @@ def create_object_to_pass_to_user_code(self, context) -> Any:  # pylint: disable
         Default behavior for new class-based resources is to return itself, passing
         the actual resource object to user code.
         """
-        return self
+        return cast(ResValue, self)
+
+    @classmethod
+    def configure_at_launch(cls, **kwargs) -> "PartialResource[ResValue]":
+        """
+        Returns a partially initialized copy of the resource, with remaining config fields
+        set at runtime.
+        """
+        return PartialResource(cls, data=kwargs)
+
+    # The following methods are used to implement the descriptor protocol
+    # https://docs.python.org/3/howto/de scriptor.html
+    #
+    # Used to adjust the types of resource inputs and outputs, e.g. resource dependencies can be passed in
+    # as PartialResources or Resources, but will always be returned as Resources
+    # Very similar to https://github.com/pydantic/pydantic/discussions/4262
+
+    def __set_name__(self, _owner, name):
+        self._assigned_name = name
+
+    def __get__(self: Self, obj: Any, __owner: Any) -> Self:
+        return cast(Self, getattr(obj, self._assigned_name))
+
+    def __set__(
+        self: Self, obj: Optional[object], value: Union[Self, "PartialResource[Self]"]
+    ) -> None:
+        setattr(obj, self._assigned_name, value)
```

Comment:
If this is truly for type-checking only, I think this another case where putting this in a class and making it a mixin will be clearer and make this more "delete-able". So like a class called `AllowPartialResourceInitParams" or something.

Reply:
Updated to a standalone mixin class in separate file